src/NUnitCommon/nunit.extensibility/ExtensionWrapper.cs

```diff
@@ -0,0 +1,105 @@
+﻿// Copyright (c) Charlie Poole, Rob Prouse and Contributors. MIT License - see LICENSE.txt
+
+using System;
+using System.Collections.Generic;
+using System.Reflection;
+
+namespace NUnit.Extensibility
+{
+    using Wrappers;
+
+    public abstract class ExtensionWrapper
+    {
+        static protected readonly Logger log = InternalTrace.GetLogger(typeof(ExtensionWrapper));
+
+        private object _wrappedInstance;
+        private Type _wrappedType;
+
+        public static ExtensionWrapper Wrap(object extension, string path)
+        {
+            switch (path)
+            {
+                // NYI and probably will not be implemented as we know of
+                // no existing v3 IService extensions.
+                //case "/NUnit/Engine/TypeExtensions/IService":
+                //    return new ServiceWrapper(extension);
+                case "/NUnit/Engine/TypeExtensions/ITestEventListener":
+                    return new TestEventListenerWrapper(extension);
+                // TODO: NYI
+                //case "/NUnit/Engine/TypeExtensions/IDriverFactory":
+                //    return new DriverFactoryWrapper(extension);
+                case "/NUnit/Engine/TypeExtensions/IProjectLoader":
+                    return new ProjectLoaderWrapper(extension);
+                case "/NUnit/Engine/TypeExtensions/IResultWriter":
+                    return new ResultWriterWrapper(extension);
+                default:
+                    throw new NUnitExtensibilityException($"No wrapper available for extension path {path}");
+            }
+        }
+
+        protected ExtensionWrapper(object wrappedInstance)
+        {
+            _wrappedInstance = wrappedInstance;
+            _wrappedType = wrappedInstance.GetType();
+        }
+
+        protected void Invoke(string methodName, params object[] args)
+        {
+            WrapMethod(methodName, Type.GetTypeArray(args)).Invoke(_wrappedInstance, args);
```

Comment:
Although elegant, there are two problem with `Type.GetTypeArray`.
It doesn't support `null` arguments as you can't get the type of `null`.
I don't see any of your wrappers allowing `null`. So it might not be relevant.
The other problem is that you get a new array every call which means that the `MethodSignature`  key is different and there is no caching. To a lesser extent it is a small hit on the GC, but with this calling reflection that would be irrelevant.
I can fix the caching by overriding `GetHashCode` and `Equals`, but as the signature is fixed at compile time, it would be better to pass the types in.

I added overloads allowing passing in the ArgTypes.
I also added overloads for the common method that have one `string` argument.